src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs

```diff
@@ -71,19 +89,194 @@ protected override void EndProcessing()
+
+                if (Top > 0 && Top < listToOutput.Count)
+                {
+                    // There may be a faster algorithm to sort while eliminating duplicates and retaining only
+                    // the top N records; for now though, this gets the job done
+                    listToOutput = listToOutput.GetRange(0, Top);
+                }
+                else if (Bottom > 0 && Bottom < listToOutput.Count)
+                {
+                    int lastItemIndex = listToOutput.Count - 1;
+                    listToOutput = listToOutput.GetRange(lastItemIndex - Bottom, lastItemIndex);
+                }
+
+                // write to output stream
+                foreach (var x in listToOutput)
+                {
+                    WriteObject(x.inputObject);
+                }
+            }
+            // Otherwise, use an indexed min-/max-heap to sort all objects
+            else
+            {
+                // The heap entry comparer will sort the items in the heap
+                var comparer = new IndexedMinMaxHeapComparer(orderByProperty.Comparer);
+
+                // Identify how many items will be in the heap
+                int indexedMinMaxHeapSize = Top > 0 ? Top : Bottom;
+
+                // A small list will store the data for the heap 
+                var indexedMinMaxHeap = new List<IndexedMinMaxHeapEntry>(indexedMinMaxHeapSize + 1);
+
+                // Identify the comparator (the value all comparisons will be made against based on whether we're
+                // doing a Top N or Bottom N sort)
+                // Note: All comparison results in the loop below are performed related to the value of the
+                // comparator. OrderByPropertyComparer.Compare will return -1 to indicate that the lhs is smaller
+                // if an ascending sort is being executed, or -1 to indicate that the lhs is larger if a descending
+                // sort is being executed. The comparator will be -1 if we're executing a Top N sort, or 1 if we're
+                // executing a Bottom N sort. These two pairs of states allow us to perform the proper comparison
+                // regardless of whether we're executing an ascending or descending Top N or Bottom N sort. This
+                // allows us to build a min-heap or max-heap for each of these sorts with the exact same logic.
+                // Min-heap: used for faster processing of a top N descending sort and a bottom N ascending sort 
+                // Max-heap: used for faster processing of a top N ascending sort and a bottom N descending sort 
+                int comparator = Top > 0 ? -1 : 1;
+
+                // Tracking the index is necessary so that unsortable items can be output at the end, in the order
+                // in which they were received.
+                int entryIndex = 0;
+                foreach (var orderMatrixEntry in orderByProperty.OrderMatrix)
```

Comment:
Funny. I used a foreach loop because I didn't originally have the index tracked, then while testing realized I needed the index so added it and didn't think to switch to for loop (which makes much more sense with the index being tracked). Fixed locally and will commit soon.